expression/builtin_compare.go

```diff
@@ -382,38 +383,94 @@ func temporalWithDateAsNumEvalType(argTp *types.FieldType) (argEvalType types.Ev
 	return
 }
 
-// GetCmpTp4MinMax gets compare type for GREATEST and LEAST and BETWEEN
-func GetCmpTp4MinMax(args []Expression) (argTp types.EvalType) {
-	datetimeFound, isAllStr := false, true
-	cmpEvalType, isStr, isTemporalWithDate := temporalWithDateAsNumEvalType(args[0].GetType())
-	if !isStr {
-		isAllStr = false
+func aggregateType(args []Expression) *types.FieldType {
+	aggType := *args[0].GetType()
+	var (
+		newUnsignedFlag   = mysql.HasUnsignedFlag(aggType.Flag)
+		mixedUnsignedFlag = false
+	)
+	for i := 1; i < len(args); i++ {
+		item := args[i].GetType()
+		mixedUnsignedFlag = mixedUnsignedFlag || (newUnsignedFlag != mysql.HasUnsignedFlag(item.Flag))
+		aggType.Tp = types.MergeFieldType(aggType.Tp, item.Tp)
+		aggType.Decimal = int(math.Max(float64(aggType.Decimal), float64(item.Decimal)))
 	}
-	if isTemporalWithDate {
-		datetimeFound = true
+
+	if mixedUnsignedFlag && types.IsTypeInteger(aggType.Tp) {
+		bumpRange := false
+		for i := range args {
+			item := args[i].GetType()
+			bumpRange = bumpRange ||
+				(mysql.HasUnsignedFlag(item.Flag) &&
+					(item.Tp == aggType.Tp || item.Tp == mysql.TypeBit))
+		}
+		if bumpRange {
+			newType := mysql.TypeUnspecified
+			switch aggType.Tp {
+			case mysql.TypeTiny:
+				newType = mysql.TypeShort
+			case mysql.TypeShort:
+				newType = mysql.TypeInt24
+			case mysql.TypeInt24:
+				newType = mysql.TypeLong
+			case mysql.TypeLong:
+				newType = mysql.TypeLonglong
+			case mysql.TypeLonglong:
+				newType = mysql.TypeNewDecimal
+			}
+			aggType.Tp = newType
+		}
 	}
-	lft := args[0].GetType()
-	for i := range args {
-		rft := args[i].GetType()
-		var tp types.EvalType
-		tp, isStr, isTemporalWithDate = temporalWithDateAsNumEvalType(rft)
-		if isTemporalWithDate {
-			datetimeFound = true
+
+	if mysql.HasUnsignedFlag(aggType.Flag) && !mixedUnsignedFlag {
+		aggType.Flag |= mysql.UnsignedFlag
+	}
+
+	return &aggType
+}
+
+// ResolveType4Between resolves eval type for between expression.
+func ResolveType4Between(args [3]Expression) types.EvalType {
+	cmpTp := args[0].GetType().EvalType()
+	for i := 1; i < 3; i++ {
+		cmpTp = getBaseCmpType(cmpTp, args[i].GetType().EvalType(), nil, nil)
+	}
+
+	hasTemporal := false
+	if cmpTp == types.ETString {
+		for _, arg := range args {
+			if types.IsTypeTemporal(arg.GetType().Tp) {
+				hasTemporal = true
+				break
+			}
 		}
-		if !isStr {
-			isAllStr = false
+		if hasTemporal {
+			cmpTp = types.ETDatetime
 		}
-		cmpEvalType = getBaseCmpType(cmpEvalType, tp, lft, rft)
-		lft = rft
-	}
-	argTp = cmpEvalType
-	if cmpEvalType.IsStringKind() {
-		argTp = types.ETString
 	}
-	if isAllStr && datetimeFound {
-		argTp = types.ETDatetime
+
+	return cmpTp
+}
+
+// ResolveType4Extremum gets compare type for GREATEST and LEAST and BETWEEN (mainly for datetime).
+func ResolveType4Extremum(args []Expression) types.EvalType {
+	aggType := aggregateType(args)
+
+	var temporalItem *types.FieldType
+	if types.ResultMergeType(aggType.Tp) == types.ETString {
```

Comment:
Useless ResultMergeType?
aggType.EvalType is enough

Reply:
`EvalType()` returns the evaluated type of the `FieldType` instead of the result merge type which will convert a `Datetime` into a string result instead of the `ETDatetime`.

Reply:
In another word, `ResultMergeType` converts all types into 4 basic result types: `ETInt`, `ETReal`, `ETDecimal`, `ETString`.

Reply:
I think we do not need this, we can check whether the `ETType == ETDatetime || ETType == ETString || ...`